### ISSUE TITLE
[Question] Unable to launch multi-nodes job using optimal_example.py

### ISSUE THREAD

**Issue by crazyguitar** [NVIDIA/nvidia-resiliency-ext#13]
Hi team,

We are attempting to launch a multi-node job using the `optimal_example.py` through the Slurm job scheduler. However, we are facing difficulties as our jobs are not running successfully. When running single-node jobs, we observed that the `inprocess.Wrapper` worked smoothly, restarting the training loop whenever a RuntimeError was encountered. According to the [documentation](https://github.com/NVIDIA/nvidia-resiliency-ext/blob/main/examples/inprocess/optimal_example.py#L28-L32), launching a training job through `torchrun` may pose some compatibility issues. We would appreciate if you could provide instructions on how to correctly launch a multi-node job with the example. 

## Environment

* AWS HyperPod Slurm cluster
* Instance type: p5.48xlarge
* GPU: NVIDIA H100 80GB HBM3 
* nvidia-resiliency-ext version: v0.2.0

## Experiment

Experiment dockerfile

```
FROM nvcr.io/nvidia/pytorch:24.10-py3

RUN git clone https://github.com/NVIDIA/nvidia-resiliency-ext.git /opt/nvidia-resiliency-ext
RUN cd /opt/nvidia-resiliency-ext && pip install .
```

Instructions to submit a Slurm job

```bash
nodes=2
salloc -N ${nodes}
srun \
  --container-image /fsx/sqsh/nvidia-resiliency-ext+latest.sqsh \
  --container-mounts /fsx:/fsx,/var/log/aws/clusters:/var/log/aws/clusters \
  torchrun \
    --nproc_per_node=8 \
    --rdzv_id=100 \
    --rdzv_backend=c10d \
    --nnodes=1 \
    --rdzv_endpoint=ip-10-0-128-155:29400 \
    /fsx/nvidia-resiliency-ext/examples/inprocess/optimal_example.py \
      --device cuda \
      --size 128 \
      --layers 8 \
      --total-iterations 10000 \
      --path /fsx/resiliency/checkpoints \
      --log-interval 10 \
      --chkpt-interval 100 \
      --fault-prob 0.001
```

Workflow

![Image](https://github.com/user-attachments/assets/3467ad64-c009-4a97-aee0-0a4ba808ea9a)

multi-node errors we got

<img width="1370" alt="Image" src="https://github.com/user-attachments/assets/9989a83d-74c6-4f2a-a6b7-6b0f1ae5856a" />

**Comment by crazyguitar** [comment]
Hi Team,

We figure out a solution to launch a multi-node job with the following patch. 

```patch
diff --git a/examples/inprocess/optimal_example.py b/examples/inprocess/optimal_example.py
index 6b60665..5bd8610 100644
--- a/examples/inprocess/optimal_example.py
+++ b/examples/inprocess/optimal_example.py
@@ -117,6 +117,19 @@ def parse_args():
     return parser.parse_args()
 
 
+import nvidia_resiliency_ext.inprocess.store
+
+class TCPStore(torch.distributed.TCPStore, nvidia_resiliency_ext.inprocess.store.StoreMixin):
+    TCP_STORE_HOST_RANK = 0
+
+    def __init__(self, *a, **kw):
+        super().__init__(*a, **kw)
+
+    @property
+    def critical_ranks(self):
+        return (self.TCP_STORE_HOST_RANK,)
+
+
 # TCPStore created by the Wrapper uses ``(MASTER_PORT + 2)`` port for the
 # internal Wrapper TCPStore to avoid conflicts with application's TCPStore
 # listening on ``(MASTER_PORT + 1)``, and with a TCPStore created by
@@ -125,7 +138,13 @@ def parse_args():
 # An instance of ``inprocess.CallWrapper` is automatically injected into
 # wrapped function arguments when Wrapper is invoked.
 @inprocess.Wrapper(
-    store_kwargs={'port': int(os.getenv('MASTER_PORT', 29500)) + 2},
+    store_factory=TCPStore,
+    store_kwargs={
+      'host_name': os.environ['MASTER_ADDR'],
+      'port': int(os.getenv('MASTER_PORT', 29500)),
+      'world_size': int(os.environ['WORLD_SIZE']),
+      'is_master': False,
+    },
     health_check=inprocess.health_check.CudaHealthCheck(),
 )
 def train(
@@ -204,7 +223,7 @@ def train(
                     torch.save(checkpoint, checkpoint_path)
 
         # Randomly trigger an example fault
-        if random.random() < args.fault_prob:
+        if rank != 0 and random.random() < args.fault_prob:
             raise_timestamp = time.perf_counter()
             raise RuntimeError(f'example fault at {iteration=} from {rank=}')
 
@@ -256,11 +275,9 @@ def main():
     # created by ``torch.distributed.run`` and listening on ``MASTER_PORT``.
     store = torch.distributed.TCPStore(
         host_name=os.environ['MASTER_ADDR'],
-        port=int(os.environ['MASTER_PORT']) + 1,
+        port=int(os.environ['MASTER_PORT']),
         world_size=int(os.environ['WORLD_SIZE']),
-        is_master=(int(os.environ['RANK']) == 0),
-        multi_tenant=True,
-        wait_for_workers=True,
+        is_master=False,
         use_libuv=True,
     )
```

From the [documentation](https://nvidia.github.io/nvidia-resiliency-ext/inprocess/usage_guide.html#restrictions):

> node failure on rank 0 causes termination of the entire job

It appears that rank 0 shouldn't fail. Do you have any suggestion how to free this restriction?

**Comment by szmigacz** [comment]
> From my understanding. it looks like you were not using torchrun or ft_launcher to launch a job
That's correct. With enroot + pyxis regular SLURM env vars are translated to PyTorch distributed env vars (`RANK`, `WORLD_SIZE`, etc.) and multi-node multi-gpu job can be launched by directly SLURM without any helper.

**Comment by crazyguitar** [comment]
gotcha. thank u for the help.
